Fixes #18 - Update login and discovery method of nodes

AntiD2ta opened the pull request:
Fixes #18 

**Changes:**

- A login method is called next to the creation of a node, for get at least a seed address. If the node is in a remote location(only for dispatcher nodes) an address of a seed node must be given. Otherwise, discovery by broadcast is done and seed nodes available in the network can reply to login attempts.
- If a node `A` is connected to a seed `S`, and `S` crashes, A will notice this and will disconnect from `S` to not send more messages to it. If a new seed `S`' appears and if is reachable, the nodes will automatically connect to `S'`. See commit details for more. Added `discoverPeer`, `getSeeds`, `findSeeds`, `ping`, `disconnectFromSeeds`. All this functions are correctly descripted in the code by comments.
- Update list of port usage in params.py
- The client's user can write a address of a seed node active in a network.txt file in project's root. This is useful if the client is not in the same subnet that the system, and he lose all links to seed nodes. In this case, if he can find an address of an active seed node, he can reconnect to the system without lose all his progress.

**Testing:**
Most cases have been successfully tested.

The thing that most require testing is this case: a remote dispatcher node started with the address of a seed:

- [ ] The node connects, but there are more actives seed nodes, the dispatcher acknowledge this, the initial node crashes, the dispatcher is still connected to the network? 
- [ ] The node connects, and there are more actives seed nodes, but the initial given seed crashes before a `GET_SEED` request is done, What happens next? What is the desired action?

AntiD2ta committed:
[seed] - taskManager's code refactored

AntiD2ta committed:
[seed] refs #18 - Pipeline for the seed nodes know each other created

Created a list property seeds for the Seed and a thread seedManager for updating the list

AntiD2ta committed:
[seed] refs #18 - Discovery by broadcast coded and tested between seed nodes.

Login and initial discovery is better placed in a login method to be called before serve, so serve is called will all the initial params ok

Side note: Broadcast not work offline.

AntiD2ta committed:
[util] refs #18 - Add broadcast default port to params

AntiD2ta committed:
[seed] refs #18 - Complete login process for seed nodes.

Works with and without connection. New args params added, execute python <node_name>.py - h for details. 
Try not to start several seed nodes in the same host, because they cannot share the same broadcast listener port, do it for developing purposes only

AntiD2ta committed:
[util] refs #18 - Upgrade and move discoverPeer and getSeeds function to utils.

Reason of the moving: To share this functions between the different types of nodes

AntiD2ta committed:
[seed] refs #18 - Do some fixes

Fixes:
- Check if Get Seeds fails to connect to the seed for some reason
- Send seeds list for GET_SEEDS request using send_pyobj instead of send_json because send_json sends a list of lists, and we need a list of tuples
- Handle PING requests

AntiD2ta committed:
[scrapper] refs #18 - Implement login and discovery mechanisms

Changes:
- Add login method
- Add thread for connect to seed nodes, especially when a new seed is discovered
- Add process to make PING request to a given seed.
- Add process to find new seed nodes from time to time

AntiD2ta committed:
[dispatcher] refs #18 - Implement login and discovery mechanisms

Changes:
- Add login method
- Automate process of connecting and disconnecting on sockets given the availability of peers. For this, a new thread is used: Disconnect to Seeds, and semaphores are used
- Use findSeeds as a process to discover seed nodes and to update list of seeds.

Testing:
The thing that most require testing is this case: a remote dispatcher node started with the address of a seed:
1-  The node connects, but there are more actives seed nodes, the dispatcher acknowledge this, the initial node crashes, the dispatcher is still connected to the network? 
2- The node connects, and there are more actives seed nodes, but the initial given seed crashes before a GET_SEED request is done, what happens next? what is the desired action?

AntiD2ta committed:
[scrapper] refs #18 - Automate process of connecting and disconnecting on sockets given the availability of peers

 For this, a new thread is used: Disconnect to Seeds, and semaphores are used

AntiD2ta committed:
[util] refs #18 - Do some changes on util files

Changes:
- Update port usage description on params.py
- Move findSeeds and ping function to util.py. Also this function were updated

AntiD2ta committed:
[scrapper] refs #18 - Change disconnectToSeeds name for disconnectFromSeeds.

e1Ru1o commented on util/utils.py:
```diff
@@ -52,13 +57,153 @@ def error(self, msg, mth=""):
+            else:
+                log.debug("New seed nodes where finded", "Find Seeds")
+
+            for s in tmp - seeds:
```
why not use here `dif` instead of `tmp - seeds`?

AntiD2ta replied:
Is the same thing. I prefer built in operators

e1Ru1o replied:
I am not talking about operators. 
`line 197: dif = tmp - seeds`
the value used in the `for` loop it was saved in `dif` previously

e1Ru1o commented on dispatcher.py:
```diff
@@ -77,14 +152,16 @@ def dispatch(self, queue):
     
     uuid = makeUuid(2**55, urls)
     d = Dispatcher(urls, uuid, args.address, args.port)
+    if not d.login(args.seed):
```
Notify what happened or ask to enter another address

AntiD2ta replied:
The login's logs notify what happened. Ask to enter another address is now solved.

e1Ru1o commented on dispatcher.py:
```diff
@@ -1,12 +1,16 @@
 log = Logger(name="Dispatcher")
 
+lockSocketReq = tLock()
+counterSocketReq = Semaphore()
```
In this state from the 'dispatch' method, messages will be sent even if it is not connected to any master, since it can be done `acquires`

e1Ru1o commented on scrapper.py:
```diff
@@ -38,21 +40,73 @@ def slave(tasks, notifications, idx):
+            log.info(f"Dispatcher disconnected of seed with address:{addr}:{port + 1})", "Disconnect from Seeds1")
+
+
+def connectToSeeds2(sock, peerQ):
```
use only a `connectToSeeds` with the appropriate parameters

AntiD2ta replied:
`connectToSeeds1` and `connectToSeeds2` are different in many ways, get notice. Logs, description, lock(remember that it is global) used and port used.

e1Ru1o replied:
both methods are exactly the same, you only need to pass some parameters

e1Ru1o commented on scrapper.py:
```diff
@@ -38,21 +40,73 @@ def slave(tasks, notifications, idx):
     
-def notifier(notifications):
+
+def disconnectFromSeeds2(sock, peerQ):
```
use only a `disconnectFromSeeds` with the appropriate parameters

AntiD2ta replied:
`disconnectToSeeds1` and `disconnectToSeeds2` are different in many ways, get notice. Logs, description, lock(remember that it is global) used and port used.

e1Ru1o replied:
the same situation of connectToSeeds

e1Ru1o commented on scrapper.py:
```diff
@@ -12,6 +11,9 @@
 lockSocketPull = tLock()
+lockSocketNotifier = tLock()
+counterSocketPull = Semaphore()
+counterSocketNotifier = Semaphore()
```
In this state both semaphores are acquirable, is that the expected behavior?

AntiD2ta committed:
[dispatcher] [scrapper] refs #18 - Fix semaphores pipeline.

Fixes:
- Semaphores are now initializated with value 0
-  Add NOBLOCK flag to socketPull.recv to release the lockSocketPull if there are no messages to be pulled either because there is no active seed nodes, or because there is no task to be push by the seed nodes.

AntiD2ta committed:
[seed] refs #18 - Pushed tasks upgraded

Upgrade:
- This task are now of type [False, counter] where counter is the number of times that tasks has been requested by a dispatcher. When counter is 10, the task is repushed and counter is set to 0.

AntiD2ta committed:
[dispatcher] refs #18 - Ask for a new seed address or for retry if login failed

AntiD2ta committed:
[makefile] - Add a dispatcher-remote rule

This is for the node attempts to connect directly to a seed created by make seed. Useful to connect quickly without doing discovery and when the node is not in the subnet of the system.

AntiD2ta committed:
[dispatcher] refs #18 - Add feature to give a address of a seed in runtime

Feature:
Add process getSeedFromFile that creates a network.txt for the user writes a seed_addr:seed_port to be connected in runtime.

AntiD2ta committed:
[gitignore] - Ignore network.txt

AntiD2ta committed:
[util] refs #18 - Upgrade findSeeds

Upgrade:
 Function now ping to all known seed nodes and get seeds of the last who ping was OK

AntiD2ta committed:
[util] refs #18 - Use dif variable instead of do tmp - seeds again

AntiD2ta committed:
[scrapper] refs #18 - Amend some logs

AntiD2ta committed:
[scrapper] refs #18 - Code refactored

Changes:
- Set a single connectToSeeds and disconnectToSeeds for the two objectives that this functions are used for
